src/library/scala/collection/TraversableLike.scala

```diff
@@ -605,22 +605,65 @@ trait TraversableLike[+A, +Repr] extends Any
    *           applied to this $coll. By default the string prefix is the
    *           simple name of the collection class $coll.
    */
-  def stringPrefix : String = {
+  def stringPrefix: String = {
+    /* This method is written in a style that avoids calling `String.split()`
+     * as well as methods of java.lang.Character that require the Unicode
+     * database information. This is mostly important for Scala.js, so that
+     * using the collection library does automatically bring java.util.regex.*
+     * and the Unicode database in the generated code.
+     *
+     * This algorithm has the additional benefit that it won't allocate
+     * anything except the result String in the common case, where the class
+     * is not an inner class (i.e., when the result contains no '.').
+     */
     val fqn = repr.getClass.getName
-    val cls = {
-      val idx1 = fqn.lastIndexOf('.' : Int)
-      if (idx1 != -1) fqn.substring(idx1 + 1) else fqn
-    }
-    val parts = cls.split('$')
-    val last = parts.length - 1
-    parts.zipWithIndex.foldLeft("") { case (z, (s, i)) =>
-      if (s.isEmpty) z
-      else if (i != last && s.forall(java.lang.Character.isDigit)) "" // drop prefix in method-local classes
-      else if (i == 0 || java.lang.Character.isUpperCase(s.charAt(0))) {
-        if (z.isEmpty) s else z + '.' + s
+
+    var pos: Int = fqn.length - 1
+
+    // Skip trailing $'s
+    while (pos != -1 && fqn.charAt(pos) == '$')
+      pos -= 1
+    if (pos == -1 || fqn.charAt(pos) == '.')
+      return ""
+
+    var result: String = ""
+    while (true) {
+      // Invariant: if we enter the loop, there is a non-empty part
+      val partEnd = pos + 1
+
+      // First skip digits
+      while (pos != -1 && fqn.charAt(pos) <= '9' && fqn.charAt(pos) >= '0')
+        pos -= 1
+      val lastNonDigit = pos
+
+      // Then skip remaining of the part
+      while (pos != -1 && fqn.charAt(pos) != '$' && fqn.charAt(pos) != '.')
+        pos -= 1
+
+      if (pos == lastNonDigit && partEnd != fqn.length) {
+        // This non-last part contains only digits -> drop prefix in method-local part
+        return result
+      }
+
+      val partStart = pos + 1
+
+      // Skip to the next part
+      while (pos != -1 && fqn.charAt(pos) == '$')
+        pos -= 1
+
+      // Handle the actual content of the part
+      val atEnd = pos == -1 || fqn.charAt(pos) == '.'
+      if (atEnd || (fqn.charAt(partStart) <= 'Z' && fqn.charAt(partStart) >= 'A')) {
```

Comment:
I didn't stop to question the original algorithm, TBH ^^ I found that a bit weird, but I thought there was a good reason for that. You were the one to write the previous version of the algorithm. Do you remember why it was only checking for uppercase? Was it some clever way to filter out `$$lambda` something?